File: backend/src/modules/producto/producto.service.ts
import {
  BadRequestException,
  Injectable,
  NotFoundException,
} from "@nestjs/common";
import { PrismaService } from "@common/config/database/prisma.service";
import { CreateProductoDto } from "./dto/create-producto.dto";
import { UpdateProductoDto } from "./dto/update-producto.dto";
import { PublicarProductoDto } from "./dto/publicar-producto.dto";
import { ActualizarProductoDto } from "./dto/actualizar-producto.dto";

@Injectable()
export class ProductoService {
  constructor(private readonly prisma: PrismaService) {}

  async publicar(dto: PublicarProductoDto, urlsImagenes: string[]) {
    if (urlsImagenes.length === 0) {
      throw new BadRequestException("Se requiere al menos una imagen");
    }

    const producto = await this.prisma.producto.create({
      data: {
        nombre: dto.nombre,
        descripcion: dto.descripcion ?? null,
        talle: dto.talle,
        precioCentavos: BigInt(dto.precioCentavos),
        stock: dto.stock ?? 0,
        imagenes: urlsImagenes,
        activo: true,
      },
    });

    return this.construirRespuesta(producto);
  }

  async crear(data: CreateProductoDto) {
    const datosPrisma = this.mapearParaPrisma(data);
    const producto = await this.prisma.producto.create({ data: datosPrisma });
    return this.construirRespuesta(producto);
  }

  async listar(params?: {
    activo?: boolean | undefined;
    talle?: string | undefined;
    q?: string | undefined;
    pagina?: number | undefined;
    tamano?: number | undefined;
  }) {
    const { activo, talle, q, pagina = 1, tamano = 20 } = params ?? {};
    const where: Record<string, unknown> = {};
    if (activo !== undefined) where.activo = activo;
    if (talle) where.talle = talle;
    if (q && q.trim()) {
      where.nombre = { contains: q.trim(), mode: 'insensitive' };
    }

    const [total, productos] = await Promise.all([
      this.prisma.producto.count({ where }),
      this.prisma.producto.findMany({
        where,
        skip: (pagina - 1) * tamano,
        take: tamano,
        orderBy: { creadoEn: "desc" },
      }),
    ]);

    return {
      productos: productos.map(this.construirRespuesta),
      total,
      pagina,
      tamano,
    };
  }

  async obtenerUno(id: string) {
    const producto = await this.prisma.producto.findUnique({ where: { id } });
    if (!producto) throw new NotFoundException("Producto no encontrado");
    return this.construirRespuesta(producto);
  }

  async actualizar(id: string, data: UpdateProductoDto) {
    const existente = await this.prisma.producto.findUnique({ where: { id } });
    if (!existente) throw new NotFoundException("Producto no encontrado");
    const datosPrisma = this.mapearUpdateParaPrisma(data);
    const producto = await this.prisma.producto.update({
      where: { id },
      data: datosPrisma,
    });
    return this.construirRespuesta(producto);
  }

  async actualizarAdmin(
    id: string,
    dto: ActualizarProductoDto,
    urlsNuevasImagenes?: string[],
  ) {
    const existente = await this.prisma.producto.findUnique({ where: { id } });
    if (!existente) throw new NotFoundException("Producto no encontrado");

    const datos: Record<string, unknown> = {};
    if (dto.nombre !== undefined) datos.nombre = dto.nombre;
    if (dto.descripcion !== undefined)
      datos.descripcion = dto.descripcion ?? null;
    if (dto.talle !== undefined) datos.talle = dto.talle;
    if (dto.precioCentavos !== undefined)
      datos.precioCentavos = BigInt(dto.precioCentavos);
    if (dto.stock !== undefined) datos.stock = dto.stock;
    if (urlsNuevasImagenes !== undefined) datos.imagenes = urlsNuevasImagenes;

    const producto = await this.prisma.producto.update({
      where: { id },
      data: datos,
    });

    return this.construirRespuesta(producto);
  }

  async desactivar(id: string) {
    const producto = await this.prisma.producto.findUnique({ where: { id } });
    if (!producto) throw new NotFoundException("Producto no encontrado");
    if (producto.activo === false)
      throw new BadRequestException("Este producto ya está desactivado");

    const productoActualizado = await this.prisma.producto.update({
      where: { id },
      data: { activo: false },
    });

    return this.construirRespuesta(productoActualizado);
  }

  async eliminar(id: string) {
    const existente = await this.prisma.producto.findUnique({ where: { id } });
    if (!existente) throw new NotFoundException("Producto no encontrado");
    await this.prisma.producto.delete({ where: { id } });
    return { id, eliminado: true };
  }

  private mapearParaPrisma(data: CreateProductoDto) {
    return {
      nombre: data.nombre,
      talle: data.talle,
      precioCentavos: BigInt(data.precioCentavos),
      stock: data.stock,
      descripcion: data.descripcion ?? null,
      imagenes: data.imagenes ?? [],
      activo: data.activo ?? true,
    };
  }

  private mapearUpdateParaPrisma(data: UpdateProductoDto) {
    const datos: Record<string, unknown> = {};
    if (data.nombre !== undefined) datos.nombre = data.nombre;
    if (data.talle !== undefined) datos.talle = data.talle;
    if (data.precioCentavos !== undefined)
      datos.precioCentavos = BigInt(data.precioCentavos);
    if (data.stock !== undefined) datos.stock = data.stock;
    if (data.descripcion !== undefined)
      datos.descripcion = data.descripcion ?? null;
    if (data.imagenes !== undefined) datos.imagenes = data.imagenes;
    if (data.activo !== undefined) datos.activo = data.activo;
    return datos;
  }

  private construirRespuesta(p: {
    id: string;
    nombre: string;
    descripcion: string | null;
    talle: string;
    precioCentavos: bigint | number;
    stock: number;
    imagenes: string[];
    activo: boolean;
    creadoEn: Date;
    actualizadoEn: Date;
  }): {
    id: string;
    nombre: string;
    descripcion: string | null;
    talle: string;
    precioCentavos: number;
    stock: number;
    imagenes: string[];
    activo: boolean;
    creadoEn: Date;
    actualizadoEn: Date;
  } {
    const precioNum: number = typeof p.precioCentavos === 'bigint'
      ? Number(p.precioCentavos)
      : p.precioCentavos;

    return {
      id: p.id,
      nombre: p.nombre,
      descripcion: p.descripcion,
      talle: p.talle,
      precioCentavos: precioNum,
      stock: p.stock,
      imagenes: p.imagenes,
      activo: p.activo,
      creadoEn: p.creadoEn,
      actualizadoEn: p.actualizadoEn,
    };
  }
}

File: backend/src/main/java/com/sistemapedidos/modules/producto/ProductoService.java
package com.sistemapedidos.modules.producto;

import com.sistemapedidos.common.error.BadRequestException;
import com.sistemapedidos.common.error.NotFoundException;
import com.sistemapedidos.domain.entity.Producto;
import com.sistemapedidos.domain.repository.ProductoRepository;
import com.sistemapedidos.modules.producto.dto.ActualizarProductoDto;
import com.sistemapedidos.modules.producto.dto.CreateProductoDto;
import com.sistemapedidos.modules.producto.dto.PaginadoProductos;
import com.sistemapedidos.modules.producto.dto.ProductoResponse;
import com.sistemapedidos.modules.producto.dto.PublicarProductoDto;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class ProductoService {

    private final ProductoRepository productoRepository;

    @Transactional
    public ProductoResponse publicar(PublicarProductoDto dto, List<String> urlsImagenes) {
        if (urlsImagenes.isEmpty()) {
            throw new BadRequestException("Se requiere al menos una imagen");
        }

        Producto producto = new Producto();
        producto.setNombre(dto.nombre());
        producto.setDescripcion(dto.descripcion());
        producto.setTalle(dto.talle());
        producto.setPrecioCentavos(dto.precioCentavos());
        producto.setStock(dto.stock() != null ? dto.stock() : 0);
        producto.setImagenes(urlsImagenes);
        producto.setActivo(true);
        productoRepository.save(producto);

        return aRespuesta(producto);
    }

    @Transactional
    public ProductoResponse crear(CreateProductoDto dto) {
        Producto producto = new Producto();
        producto.setNombre(dto.nombre());
        producto.setTalle(dto.talle());
        producto.setPrecioCentavos(dto.precioCentavos());
        producto.setStock(dto.stock());
        producto.setDescripcion(dto.descripcion());
        producto.setImagenes(dto.imagenes() != null ? dto.imagenes() : List.of());
        producto.setActivo(dto.activo() != null ? dto.activo() : true);
        productoRepository.save(producto);

        return aRespuesta(producto);
    }

    @Transactional(readOnly = true)
    public PaginadoProductos listar(Boolean activo, String talle, String q, int pagina, int tamano) {
        int paginaSegura = Math.max(pagina, 1);
        int tamanoSeguro = Math.max(tamano, 1);

        Page<Producto> page = productoRepository.findAll(
                ProductoSpecifications.filtros(activo, talle, q),
                PageRequest.of(paginaSegura - 1, tamanoSeguro, Sort.by(Sort.Direction.DESC, "creadoEn")));

        List<ProductoResponse> productos = page.getContent().stream()
                .map(this::aRespuesta)
                .toList();

        return new PaginadoProductos(productos, page.getTotalElements(), paginaSegura, tamanoSeguro);
    }

    @Transactional(readOnly = true)
    public ProductoResponse obtenerUno(String id) {
        return aRespuesta(obtenerEntidad(id));
    }

    @Transactional
    public ProductoResponse actualizarAdmin(String id, ActualizarProductoDto dto, List<String> urlsNuevasImagenes) {
        Producto producto = obtenerEntidad(id);

        if (dto.nombre() != null) producto.setNombre(dto.nombre());
        if (dto.descripcion() != null) producto.setDescripcion(dto.descripcion());
        if (dto.talle() != null) producto.setTalle(dto.talle());
        if (dto.precioCentavos() != null) producto.setPrecioCentavos(dto.precioCentavos());
        if (dto.stock() != null) producto.setStock(dto.stock());
        if (urlsNuevasImagenes != null) producto.setImagenes(urlsNuevasImagenes);

        productoRepository.save(producto);
        return aRespuesta(producto);
    }

    @Transactional
    public ProductoResponse desactivar(String id) {
        Producto producto = obtenerEntidad(id);
        if (!producto.getActivo()) {
            throw new BadRequestException("Este producto ya esta desactivado");
        }
        producto.setActivo(false);
        productoRepository.save(producto);
        return aRespuesta(producto);
    }

    @Transactional
    public void eliminar(String id) {
        Producto producto = obtenerEntidad(id);
        productoRepository.delete(producto);
    }

    public Producto obtenerEntidad(String id) {
        return productoRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("Producto no encontrado"));
    }

    private ProductoResponse aRespuesta(Producto p) {
        return new ProductoResponse(
                p.getId(),
                p.getNombre(),
                p.getDescripcion(),
                p.getTalle(),
                p.getPrecioCentavos(),
                p.getStock(),
                p.getImagenes(),
                p.getActivo(),
                p.getCreadoEn(),
                p.getActualizadoEn());
    }
}
